source/daq/egg3_reader.cc

```diff
@@ -0,0 +1,185 @@
+            f_start_paused( true ),
+            f_paused( true )
+    {
+    }
+
+    egg3_reader::~egg3_reader()
+    {
+        cleanup_file();
+    }
+
+    void egg3_reader::initialize()
+    {
+        f_paused = f_start_paused;
+
+        out_buffer< 0 >().initialize( f_length );
+
+        LDEBUG( plog, "opening egg file [" << f_egg_path << "]" );
+        f_egg = monarch3::Monarch3::OpenForReading( f_egg_path );
+        f_egg->ReadHeader();
+        // do we want/need to do anything with the header?
+        //const monarch3::M3Header *t_egg_header = f_egg->GetHeader();
+        return;
+
+    }
+
+    void egg3_reader::execute( midge::diptera* a_midge )
+    {
+        try
+        {
+            LDEBUG( plog, "Executing the egg3_reader" );
+            //TODO  use header to loop streams so we can send more than one?
+            //const monarch3::M3Header *t_egg_header = f_egg->GetHeader();
+            const monarch3::M3Stream* t_stream = f_egg->GetStream( 0 );
+            const monarch3::M3Record* t_record = t_stream->GetChannelRecord( 0 );
+
+            time_data* t_data = nullptr;
+
+            // starting not in a paused state is not currently known to work
+            if ( !f_paused ) {
+                if( ! out_stream< 0 >().set( stream::s_start ) ) return;
+            }
+
+            // starting execution loop
+            while (! is_canceled() )
+            {
+                if( (out_stream< 0 >().get() == stream::s_stop) )
+                {
+                    LWARN( plog, "Output stream(s) have stop condition" );
+                    break;
+                }
+                if( have_instruction() )
+                {
+                    if( f_paused && use_instruction() == midge::instruction::resume )
+                    {
+                        LDEBUG( plog, "egg reader resuming" );
+                        if( ! out_stream< 0 >().set( stream::s_start ) ) throw midge::node_nonfatal_error() << "Stream 0 error while starting";
+                        f_paused = false;
+                    }
+                    else if ( !f_paused && use_instruction() == midge::instruction::pause )
+                    {
+                        LDEBUG( plog, "egg reader pausing" );
+                        if( ! out_stream< 0 >().set( stream::s_stop ) ) throw midge::node_nonfatal_error() << "Stream 0 error while stopping";
+                        f_paused = true;
+                    }
+                }
+                // only read if not paused:
+                if ( ! f_paused ) {
+                    LDEBUG( plog, "not paused, reading slice" );
+                    if ( !read_slice(t_data, t_stream, t_record) ) break;
+                } else {
+                    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                }
+            }
+
+
+            // reading data goes herea
+            //TODO we probably do want support for multiple streams
+        }
+        catch(...)
```

Comment:
Would be better to catch a specific exception, or std::exception at the most general.  I don't think there's a chance of having something that doesn't inherit from std::exception here.

Reply:
This was a direct copy for tf_roach_receiver. Changed both to `catch( std::exception )`.